user/src/com/google/gwt/user/server/rpc/impl/ServerSerializationStreamWriter.java

```diff
@@ -44,73 +47,99 @@ public final class ServerSerializationStreamWriter extends
+    private final Writer writer;
+    /**
+     * A buffer used for versions before
+     * {@link AbstractSerializationStream#SERIALIZATION_STREAM_FORWARD_STREAMING_VERSION}
+     * which write the payload tokens in reverse order; {@code null} for versions
+     * that use forward streaming.
+     */
+    private final List<CharSequence> tokensInReverseOrder;
     private boolean needsComma = false;
-    private int total = 0;
-    private boolean javascript = false;
 
-    public LengthConstrainedArray() {
-      buffer = new StringBuffer();
+    public TokenEncoder(int version, Writer writer) {
+      this.writer = writer;
+      if (version < AbstractSerializationStream.SERIALIZATION_STREAM_FORWARD_STREAMING_VERSION) {
+        tokensInReverseOrder = new LinkedList<>();
+      } else {
+        tokensInReverseOrder = null;
+      }
+      try {
+        writer.append("[");
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
     }
 
-    public LengthConstrainedArray(int capacityGuess) {
-      buffer = new StringBuffer(capacityGuess);
+    /**
+     * For versions prior to {@link AbstractSerializationStream#SERIALIZATION_STREAM_FORWARD_STREAMING_VERSION}
+     * payload tokens are to be written in reverse order and therefore will be buffered in {@link #tokensInReverseOrder}
+     * For versions starting with {@link AbstractSerializationStream#SERIALIZATION_STREAM_FORWARD_STREAMING_VERSION}
+     * also the payload tokens will be sent straight into the writer instead.
+     */
+    public void addPayloadToken(CharSequence token) {
+      if (tokensInReverseOrder == null) {
+        // append token directly to writer in forward order
+        addTokenToWriter(token);
+      } else {
+        // enqueue token in reverse order buffer for later writing in finish()
+        tokensInReverseOrder.add(0, token);
+      }
     }
-
+    
+    /**
+     * Appends a token at the end of the output written
+     */
     public void addToken(CharSequence token) {
-      total++;
-      if (count++ == MAXIMUM_ARRAY_LENGTH) {
-        if (total == MAXIMUM_ARRAY_LENGTH + 1) {
-          buffer.append(PRELUDE);
-          javascript = true;
+        // append token directly to writer in forward order
+        addTokenToWriter(token);
+    }
+    
+    private void addTokenToWriter(CharSequence token) {
+      try {
+        if (needsComma) {
+          writer.append(",");
         } else {
-          buffer.append("],[");
+          needsComma = true;
         }
-        count = 0;
-        needsComma = false;
-      }
-
-      if (needsComma) {
-        buffer.append(",");
-      } else {
-        needsComma = true;
+        writer.append(token);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
       }
-
-      buffer.append(token);
     }
 
-    public void addEscapedToken(String token) {
-      addToken(escapeString(token, true, this));
+    public void addEscapedToken(String token) throws IOException {
```

Comment:
adding a thrown exception is (... or should be) an api change, since calling code that doesn't catch it will now not compile.

maybe UncheckedIOException? or find another way to fail in these cases?

Reply:
Good point. It was using a RuntimeException wrapper already, but in 9fef84f3ceb273fd1f14462ac32938a2f81a4dea I've changed to UncheckedIOException and removed the "throws Exception" from the addToken methods.